File: app/src/main/java/com/example/take_out/viewmodels/OrderModel.java
package com.example.take_out.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.example.take_out.data.MyOrder;
import com.example.take_out.service.ApiResult;
import com.example.take_out.service.ResultCode;
import com.example.take_out.service.Service;

import java.util.ArrayList;
import java.util.List;

public class OrderModel extends ViewModel {
    private MutableLiveData<ApiResult<List<MyOrder>>> _orderList;

    private MutableLiveData<ApiResult<List<MyOrder>>> _getOrderList() {
        if (_orderList == null) {
            _orderList = Service.INSTANCE.getMyOrderService().getMyOrderByUserId(22);
        }
        return _orderList;
    }

    public LiveData<List<MyOrder>> getOrderList() {
        return Transformations.map(_getOrderList(), it -> {
            List<MyOrder> data = it.getData();
            if (data == null)
                return new ArrayList<>();
            else return data;
        });
    }

    public LiveData<Boolean> getReady() {
        return Transformations.map(_getOrderList(),
                it -> it.getCode() == ResultCode.SUCCESS.getCode());
    }
}

File: app/src/main/java/com/example/take_out/viewmodels/OrderModel.kt
package com.example.take_out.viewmodels

import androidx.lifecycle.*
import com.example.take_out.data.MyOrder
import com.example.take_out.service.ApiResult
import com.example.take_out.service.Service

class OrderModel(private var userId: Int) : ViewModel() {
    private val refresh = MutableLiveData(true)
    private val _orderList: LiveData<ApiResult<List<MyOrder>>> = Transformations.switchMap(refresh) {
        Service.myOrderService.getMyOrderByUserId(userId)
    }
    val loading = MutableLiveData<Boolean>()
    val orderList: LiveData<List<MyOrder>>
        get() = Transformations.map(_orderList) {
            loading.value = false
            it.data ?: listOf()
        }

    fun setUserId(userId: Int) {
        this.userId = userId
        refreshData()
    }

    fun refreshData() {
        refresh.value = true
        loading.value = true
    }

    class ViewModelFactory(private val userId: Int) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel?> create(modelClass: Class<T>): T {
            return OrderModel(userId) as T
        }
    }
}
